src/NuGet.Clients/NuGet.CommandLine/Common/ProtocolDiagnosticsJsonLogger.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Globalization;
+using System.IO;
+using Newtonsoft.Json;
+using NuGet.Protocol.Utility;
+
+namespace NuGet.CommandLine
+{
+    internal sealed class ProtocolDiagnosticsJsonLogger : IDisposable
+    {
+        private object _lock;
+        private StreamWriter _streamWriter;
+        private JsonWriter _jsonWriter;
+
+        internal ProtocolDiagnosticsJsonLogger(string fileName)
+        {
+            _lock = new object();
+            _streamWriter = new StreamWriter(fileName);
+            _jsonWriter = new JsonTextWriter(_streamWriter);
+            _jsonWriter.Formatting = Formatting.Indented;
+            _jsonWriter.WriteStartArray();
+        }
+
+        internal void OnEvent(ProtocolDiagnosticEvent pde)
+        {
+            if (_lock == null)
+            {
+                throw new ObjectDisposedException(nameof(ProtocolDiagnosticsJsonLogger));
+            }
+
+            var timestamp = pde.Timestamp.ToString("o", CultureInfo.InvariantCulture);
+
+            lock (_lock)
+            {
+                _jsonWriter.WriteStartObject();
+
+                _jsonWriter.WritePropertyName("timestamp");
+                _jsonWriter.WriteValue(timestamp);
+
+                _jsonWriter.WritePropertyName("source");
+                _jsonWriter.WriteValue(pde.Source);
+
+                _jsonWriter.WritePropertyName("url");
+                _jsonWriter.WriteValue(pde.Url);
+
+                if (pde.HeaderDuration.HasValue)
+                {
+                    _jsonWriter.WritePropertyName("headerDuration");
+                    _jsonWriter.WriteValue(pde.HeaderDuration.Value.TotalMilliseconds);
+                }
+
+                _jsonWriter.WritePropertyName("duration");
+                _jsonWriter.WriteValue(pde.EventDuration.TotalMilliseconds);
+
+                if (pde.HttpStatusCode.HasValue)
+                {
+                    _jsonWriter.WritePropertyName("httpStatusCode");
+                    _jsonWriter.WriteValue(pde.HttpStatusCode.Value);
+                }
+
+                _jsonWriter.WritePropertyName("bytes");
```

Comment:
is this bytes across the wire? response body bytes? what about `Content-Encoding` like `gzip`? Consider the Newtonsoft.Json registration index.json which is 263K non-gzipped and 8K gzipped. Looks like you built a special stream to find this so consider also include `Content-Length` if present.

Reply:
It's decompressed response body bytes.

`Content-Length` isn't always available, particularly on dynamic web applications. It should be available for nuget.org because of your architecture decisions, but I expect BaGet, and possibly even Azure Artifacts to not provide Content-Length on json files. Plus if a download times-out or otherwise fails before downloading the entire response, then it will report the "incorrect" information.

To take into account content-encoding introduces too much tech-debt. We'd have to configure the HttpClientHandler to not decompress, then use the byte counting stream on the "raw" response stream, then duplicate the http client handler's code to detect encoding and wrap a decoding stream. I've already [created](https://github.com/dotnet/corefx/issues/41408) [two](https://github.com/dotnet/corefx/issues/41410) issues on corefx which would make it easy for us to get that information, but who knows when or even if it'll be added. The benefit of measuring this isn't worth it.

Without copying the entirety of .NET Core's `SocketHttpHandler` into NuGet and modifying it as needed, I don't see any way at all to get the header bytes. I also have [another request](https://github.com/dotnet/corefx/issues/41412) on corefx to make it possible to monitor header bytes.

So, although I would love to measure over-the-wire bytes, it's infeasible to do so.

Reply:
`Content-Length` can be opportunistic and `Content-Encoding` is still visible even when getting the decoded bytes, right? Totally agreed that upleveling the decompressing is way too much.

Regarding dynamic web applications you're talking about transfer encoding `chunked` right?

Reply:
No, I mean if you create an ASP.NET Core MVC app, return some JSON from some endpoint, it will not provide a Content-Length header. So, if a NuGet server generates json by querying data from a database and dynamically creating JSON, rather than loading some pre-created JSON from some storage, there's a reasonable probability that no Content-Length header will be returned.

Reply:
Oh I didn't know clients behave well when there is no transfer encoding and no content length. I wonder when they know to stop reading on keep alive connections. 🤷‍♂

Reply:
Doing a quick search, it looks like chunked encoding might be used in HTTP 1.1 responses where the full content-length isn't known at header generation time. But if that's the case, it appears to be more magic hidden behind HttpClientHandler. I feel confident that I've seen non-compressed (encoded) responses without a content length header from HttpClient API's point of view.

Reply:
Okay, seems like the level of detail you are gathering meets your needs then. I was thinking in broader terms for more general diagnostics. My bad.